### ISSUE TITLE
Fitness functions

### ISSUE THREAD

**Issue by schrum2** [schrum2/EvoCraft-SCOPE#19]
Interactive evolution is fun, but we also need ways to automate the process a bit. Before we get too deep in this, we need some kind of basic framework for fitness functions. Specifically, a fitness function will take a shape and return a number where high numbers are better than low numbers, and indicate how good the solution is.

Make a module called `fitness_functions.py` and then define `type_count` to take the z/y/z-range values and the x/y/z-start values, an index of a member of the population (0, 1, 2, ...), and a block type. The function loops through all the voxels in the space for that particular index and counts the number of occurrences of the specified block type.

Once this simple function works, we'll need to set up a way to automate evolution using this function and verify it works, before eventually moving on to better fitness functions.

**Comment by schrum2** [comment]
Once `type_count` works, make a function in `fitness_functions.py` with this signature:
```
def eval_fitness(genomes, config, fitness):
```

This function is meant to have similarities with the `eval_fitness` method inside of the `MinecraftBreeder` class of `evolve_CPPN.py`. The `genomes` and `config` parameters are the same. The `fitness` parameter is meant to be a function like `type_count` or any other fitness function we define in a similar way.

This new `eval_fitness`  function should generate the shapes (more on this in a moment), then apply `fitness` to each member of the population and assign the resulting fitness values to the genomes accordingly.

As for as generating the shapes, we want to avoid code duplication. Also, it looks like we may end up with several different ways of generating shapes (controlled by command line parameters). Look for another issue related to this that has to be completed before this one.

**Comment by schrum2** [comment]
Correction to what the parameters to `type_count` are. Take a `dict` called `position_information` that contains the `xrange`, `starty`, etc (see #26 ) and a parameter `corner` that is a three-tuple (int,int,int) containing the minimal x/y/z coordinates for a specific shape being generated (instead of a population index)

**Comment by alejmedinajr** [comment]
I think I have completed the `type_count` function, just need to test it #26 is ready.

**Comment by schrum2** [comment]
Create `fitness_cppn_evolution.py`. This will be similar in some ways to `interactive_cppn_evolution.py`, except it does fitness based evolution. Although the code should be similar in structure to `interactive_cppn_evolution.py` it should avoid repeating it. This is part of why we are moving so much code out into `minecraft_structures.py` so it can be used in multiple places. You might also need to work with @richeyme a bit on issue #25 because you should not be copying the code for creating shapes from the CPPNs. That also needs to be moved to a separate module.

In any case, the end goal is to evolve shapes in an automated fashion using NEAT and your `type_count` fitness function. This code shows a very basic example of fitness based evolution with NEAT:
https://github.com/CodeReclaimers/neat-python/blob/master/examples/xor/evolve-minimal.py

You'll need to make a new NEAT config file and define something like the `eval_genomes` function in the example code, which is analogous to the `eval_fitness` method in our code.

**Comment by alejmedinajr** [comment]
There seems to be a problem with fitness. The shapes are remaining the same (no new shapes are appearing), and there is no improvement in the fitness.

**Comment by alejmedinajr** [comment]
The current fitness function is a good starting point, but there are some limitations such as: the fitness threshold has to be hard coded depending on the specific fitness function used, the number of generations is also hard coded, and the desired block has to also be hard coded.
